app/controllers/api/physical_switches_controller.rb

```diff
@@ -12,8 +12,59 @@ def refresh_resource(type, id, _data = nil)
       end
     end
 
+    def restart_resource(type, id, _data = nil)
+      perform_action(:restart, type, id)
+    end
+
     private
 
+    def perform_action(action, type, id)
+      if single_resource?
+        enqueue_action_single_resource(action, type, id)
+      else
+        enqueue_action_multiple_resources(action, type, id)
+      end
+    end
+
+    #
+    # Enqueues the action for a single resource.
+    #
+    # @param [symbol] action - action to be enqueued
+    # @param [symbol] type   - type of the resource
+    # @param [number] id     - id of the resource
+    #
+    def enqueue_action_single_resource(action, type, id)
+      raise BadRequestError, "Must specify an id for changing a #{type} resource" unless id
+
+      physical_switch = resource_search(id, type, collection_class(type))
+
+      api_action(type, id) do
+        begin
+          desc = "Performing #{action} for #{physical_switch_ident(physical_switch)}"
+          api_log_info(desc)
+          task_id = queue_object_action(physical_switch, desc, :method_name => action, :role => :ems_operations)
+          action_result(true, desc, :task_id => task_id)
+        rescue StandardError => err
+          action_result(false, err.to_s)
+        end
+      end
+    end
+
+    #
+    # Enqueues the action for multiple resources.
+    #   For multiple resources, when an error occurs, the error messages must
+    #   be built individually for each resource. Always responding with status 200.
```

Comment:
@douglasgabrieIt's not obvious what the difference is between `enqueue_action_single_resource` and `enqueue_action_multiple_resources`. Can you provide some more details?

Reply:
Sure! Basically a single resource request failure will look like:
```
[status = 404]
{
    "error": {
        "kind": "not_found",
        "message": "Couldn't find PhysicalSwitch with 'id'=4 [WHERE \"switches\".\"type\" IN ('PhysicalSwitch', 'ManageIQ::Providers::Lenovo::PhysicalInfraManager::PhysicalSwitch')]",
        "klass": "ActiveRecord::RecordNotFound"
    }
}
```
However, a multiple resource request failure will look like:
```
[status = 200]
{
    "results": [
        {
            "success": true,
            "message": "Performing restart for Physical Switch id:1 name: 'Lenovo Switch'",
            "task_id": "2",
            "task_href": "http://localhost:3000/api/tasks/2",
            "href": "http://localhost:3000/api/physical_switches/1"
        },
        {
            "success": false,
            "message": "Couldn't find PhysicalSwitch with 'id'=9999 [WHERE \"switches\".\"type\" IN ('PhysicalSwitch', 'ManageIQ::Providers::Lenovo::PhysicalInfraManager::PhysicalSwitch')]"
        }
    ]
}
```
The main difference here is that for a single resource request, the entire response fails, see the status.

But, for multiple resources, the failures are self contained, once some resources could execute the action successfully and others could fail, see the status again.

This is why the `enqueue_action_multiple_resources` method has a `rescue` for these errors and inside it a graceful message is mounted for the user, which doesn't occurs in `enqueue_action_single_resource`.

It made the things more clear for you @gtanzillo ? Maybe add those examples in the method documentation would be a good option, what do you think?

Reply:
One at time